### ISSUE TITLE
Crash at startup

### ISSUE THREAD

**Issue by ptytb** [LightDestory/Visual_Studio_2017_Offline_Customizer#1]
It does not work:

Faulting application name: Visual.Studio.2017.Offline.Layout.Customizer.exe, version: 1.2.3.0, time stamp: 0x5900de6c
Faulting module name: KERNELBASE.dll, version: 10.0.14393.479, time stamp: 0x582588e6
Exception code: 0xe0434352
Fault offset: 0x0000000000017788
Faulting process id: 0x171c
Faulting application start time: 0x01d2c2812970abb2
Faulting application path: C:\Users\JDoe\Downloads\Visual.Studio.2017.Offline.Layout.Customizer.exe
Faulting module path: C:\WINDOWS\System32\KERNELBASE.dll
Report Id: ea91e084-bc97-4751-b2c7-2c41471aea02
Faulting package full name: 
Faulting package-relative application ID: 

Application: Visual.Studio.2017.Offline.Layout.Customizer.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.Net.Sockets.SocketException
   at System.Net.Sockets.Socket.DoConnect(System.Net.EndPoint, System.Net.SocketAddress)
   at System.Net.ServicePoint.ConnectSocketInternal(Boolean, System.Net.Sockets.Socket, System.Net.Sockets.Socket, System.Net.Sockets.Socket ByRef, System.Net.IPAddress ByRef, ConnectSocketState, System.IAsyncResult, System.Exception ByRef)

Exception Info: System.Net.WebException
   at System.Net.WebClient.DownloadDataInternal(System.Uri, System.Net.WebRequest ByRef)
   at System.Net.WebClient.DownloadString(System.Uri)
   at VS2017OfflineCustomizer.DataContainer.GetLatestFile(System.String)
   at VS2017OfflineCustomizer.DataContainer..cctor()

Exception Info: System.TypeInitializationException
   at VS2017OfflineCustomizer.DataContainer.GetData(System.String)
   at VS2017OfflineCustomizer.Customizer..ctor(System.String)
   at VS2017OfflineCustomizer.Main..ctor()
   at VS2017OfflineCustomizer.Program.Main()

**Comment by LightDestory** [comment]
The tool requires a internet connection to work. The error occur when trying to get the latest link of vs install .exe from web ("GetLatestFile()").
There weren't any warning message because I made it a bit bad, now I moved the method inside "preinit".
So:
1) You runned the tool without a internet connection
2) Your network uses a proxy

**Comment by ptytb** [comment]
No proxy, no firewall.

Let me guess the matter is here:
```
        private static string[,] files =
        {
            {"vs_Community.exe", GetLatestFile("https://www.visualstudio.com/it/thank-you-downloading-visual-studio/?sku=Community")},
            {"vs_Professional.exe",GetLatestFile("https://www.visualstudio.com/it/thank-you-downloading-visual-studio/?sku=Professional")},
            {"vs_Enterprise.exe",GetLatestFile("https://www.visualstudio.com/it/thank-you-downloading-visual-studio/?sku=Enterprise")},
        };
```
Very bad I don't have Studio or BuildTools to fix or test it myself.

**Comment by LightDestory** [comment]
For me it is working well. I will check more.
Try using this https://db.tt/NGiJSRFeSO (version with getlatestfile() moved into preinit)

**Comment by ptytb** [comment]
Now it works! https://db.tt/NGiJSRFeSO

**Comment by LightDestory** [comment]
Before tell that you need to select a edition and click the button. I didn't nothing, just moved a method call

**Comment by LightDestory** [comment]
Now Community and Enterprise are bugged (download only 117KB) XD. Just editing some code now problems :V

**Comment by ptytb** [comment]
You are right, still blames network. Tried wget and chrome on https://www.visualstudio.com/it/thank-you-downloading-visual-studio/?sku=Community, works perfectly. I have no idea(

**Comment by LightDestory** [comment]
Try these:
1) Same as before with fixed index error: https://db.tt/k6ZM3O3hrA
2) No GetLatestFile(), direct download link: https://db.tt/o2QRfJDeOS
And if it still doesn't work, go to "Info" and check for update, if it says "network unavaible" there is something that blocks connection on your pc
If check for update works, the problem can be with the directory of download. Maybe try running as Admin. (Sorry for this problem, I am just a student and learning a bit of C# D: ). Please tell me OS, I will try to replicate on VM.

**Comment by ptytb** [comment]
Update check okay, download still fail.

**Comment by LightDestory** [comment]
Your OS?

**Comment by ptytb** [comment]
Win 10 Pro build 1607. I guess my problem happens after lots of surgery on telemetry and locking other 'redundancies'

**Comment by LightDestory** [comment]
I am also on Win10 1607 x64.
I will try to fix later. (You are the only one that provided a feedback, no other testing people and on all my PC it works).
Here a workaround to use the tool: https://www.dropbox.com/s/3g0pokgl332ok9x/vs_files.zip?dl=0
extract the file on the same folder of .exe and run the tool. You will bypass the download file. My tool just run the vs_EDITION.exe with arguments. So any problem after is not retaled with me XD (at least)
Let me know!

**Comment by ptytb** [comment]
Okay, I figured out.
```powershell
#powershell 
[Net.ServicePointManager]::SecurityProtocol = [System.Net.SecurityProtocolType]::Tls12
(New-Object System.Net.WebClient).DownloadFile("http://www.visualstudio.com/", "c:\downloads\test.txt")
```
says `The request was aborted: Could not create SSL/TLS secure channel.` even if I remove 's' from https. But other sites can be requested for me this way even with https.

The reason is Microsoft uses curve 'Nist P256' for TLS key exchange on their site. Since P256 is deprecated, . I manually disabled it, but had to turn it on.

So now your app works fine.

**Comment by LightDestory** [comment]
I am a bit embarrassed I couldnìt do anything. Thanks very much, you saved me a lot of time creating VMs and testing download. I don't know how "fix" or make it work on your previous circumstances, any help that make this tool better and allow me to learn is really appreciated.
**MAYBE** I found a fix: Switch 4.0 .net to 4.5 .net . I will make this on new release.
